### PR TITLE
feat(off-policy): add DDPG, TD3 SAC

### PR DESCRIPTION
**Pull request by Gaiejj**
## Description

Support off policy algorithms. DDPG and TD3

## Types of changes

What types of changes does your code introduce? Put an `x` in all the boxes that apply:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds core functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (update in the documentation)

## Checklist

Go over all the following points, and put an `x` in all the boxes that apply.
If you are unsure about any of these, don't hesitate to ask. We are here to help!

- [x] I have read the [CONTRIBUTION](https://omnisafe.readthedocs.io/en/latest/developer/contributing.html) guide. (**required**)
- [x] My change requires a change to the documentation.
- [ ] I have updated the tests accordingly. (*required for a bug fix or a new feature*)
- [x] I have updated the documentation accordingly.
- [ ] I have reformatted the code using `make format`. (**required**)
- [x] I have checked the code using `make lint`. (**required**)
- [ ] I have ensured `make test` pass. (**required**)

### CODE REVIEW

**Comment by rockmagma02** [comment]
please don't put [x] in not-completed items (like `make lint', `make test') in checklist

File: omnisafe/algorithms/off_policy/ddpg.py [hunk 1]
```diff
@@ -0,0 +1,342 @@
...
+    def _init(self) -> None:
+        self._buf = VectorOffPolicyBuffer(
+            obs_space=self._env.observation_space,
+            act_space=self._env.action_space,
+            size=self._cfgs.replay_buffer_cfgs.size,
+            batch_size=self._cfgs.replay_buffer_cfgs.batch_size,
+            num_envs=self._cfgs.num_envs,
+            device=self._device,
+        )
+        self._target_actor_critic = deepcopy(self._actor_critic)
```
Comment (rockmagma02) [inline omnisafe/algorithms/off_policy/ddpg.py:78]: model should init in `_init_model()`

File: omnisafe/algorithms/off_policy/ddpg.py [hunk 2]
```diff
@@ -0,0 +1,342 @@
...
+    def _init_log(self) -> None:
+        self._logger = Logger(
+            output_dir=self._cfgs.data_dir,
+            exp_name=self._cfgs.exp_name,
+            seed=self._cfgs.seed,
+            use_tensorboard=self._cfgs.use_tensorboard,
+            use_wandb=self._cfgs.use_wandb,
+            config=self._cfgs,
+        )
+        if self._cfgs.obs_normalize:
```
Comment (rockmagma02) [inline omnisafe/algorithms/off_policy/ddpg.py:96]:
```python
        what_to_save = {'pi': self._actor_critic.actor}
        if self._cfgs.obs_normalize:
            what_to_save['obs_normalizer']: obs_normalizer
```

File: omnisafe/algorithms/off_policy/ddpg.py [hunk 3]
```diff
@@ -0,0 +1,342 @@
...
+        """This is main function for algorithm update, divided into the following steps:
+
+        - :meth:`rollout`: collect interactive data from environment.
+        - :meth:`update`: perform actor/critic updates.
+        - :meth:`log`: epoch/update information for visualization and terminal log print.
+        """
+        self._logger.log('INFO: Start training')
+        start_time = time.time()
+        step = 0
+        while step < int(self._total_steps):
```
Comment (rockmagma02) [inline omnisafe/algorithms/off_policy/ddpg.py:144]: please `for` loop

File: omnisafe/algorithms/off_policy/ddpg.py [hunk 4]
```diff
@@ -0,0 +1,342 @@
...
+        act: torch.Tensor,
+        reward: torch.Tensor,
+        done: torch.Tensor,
+        next_obs: torch.Tensor,
+    ) -> None:
+        with torch.no_grad():
+            next_action = self._target_actor_critic.actor.predict(next_obs, deterministic=True)
+            next_q_value = self._target_actor_critic.reward_critic(next_obs, next_action)[0]
+            target_q_value = reward + self._cfgs.gamma * (1 - done) * next_q_value
+        q_value = self._actor_critic.reward_critic(obs, act)[0]
```
Comment (rockmagma02) [inline omnisafe/algorithms/off_policy/ddpg.py:235]: use `q_value_r`

File: omnisafe/common/buffer/offpolicy_buffer.py [hunk 5]
```diff
@@ -68,3 +72,30 @@ def sample_batch(self) -> Dict[str, torch.Tensor]:
         """Sample a batch of data from the buffer."""
         idxs = torch.randint(0, self._size, (self._batch_size,))
         return {key: value[idxs] for key, value in self.data.items()}
+
+    def sample_n_step_batch(self, n_step: int) -> Tuple[Dict[str, torch.Tensor], Any]:
```
Comment (rockmagma02) [inline omnisafe/common/buffer/offpolicy_buffer.py:76]: please add it in future version,

File: omnisafe/algorithms/__init__.py [hunk 6]
```diff
@@ -48,6 +49,7 @@
 ALGORITHMS = {
     # 'off-policy': tuple(off_policy.__all__),
```
Comment (zmsn-2077) [inline omnisafe/algorithms/__init__.py:50]: delete

File: omnisafe/algorithms/off_policy/sac.py [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+            act (torch.Tensor): current action
+            cost (torch.Tensor): current cost
+            done (torch.Tensor): current done signal
+            next_obs (torch.Tensor): next observation
+
+        Returns:
+            None
+        """
+        with torch.no_grad():
+            # Set the update noise and noise clip.
```
Comment (zmsn-2077) [inline omnisafe/algorithms/off_policy/sac.py:170]: set

File: omnisafe/algorithms/off_policy/td3.py [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+            act (torch.Tensor): current action
+            reward (torch.Tensor): current reward
+            done (torch.Tensor): current done signal
+            next_obs (torch.Tensor): next observation
+
+        Returns:
+            None
+        """
+        with torch.no_grad():
+            # Set the update noise and noise clip.
```
Comment (zmsn-2077) [inline omnisafe/algorithms/off_policy/td3.py:71]: set

File: omnisafe/algorithms/off_policy/td3.py [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+            act (torch.Tensor): current action
+            cost (torch.Tensor): current cost
+            done (torch.Tensor): current done signal
+            next_obs (torch.Tensor): next observation
+
+        Returns:
+            None
+        """
+        with torch.no_grad():
+            # Set the update noise and noise clip.
```
Comment (zmsn-2077) [inline omnisafe/algorithms/off_policy/td3.py:127]: set

File: omnisafe/adapter/offpolicy_adapter.py [hunk 10]
```diff
@@ -0,0 +1,123 @@
+# Copyright 2022-2023 OmniSafe Team. All Rights Reserved.
```
Comment (zmsn-2077) [inline omnisafe/adapter/offpolicy_adapter.py:1]: 2023

**Review by zmsn-2077 — APPROVED** [review]
LGTM.

**Pull request merged**